heidrun/nypl_mods.rb

```diff
@@ -0,0 +1,288 @@
+             .first_value.values[0]
+
+  begin_date = last_origin_info.field(datefields.join('|'))
+               .match_attribute(:point, 'start')
+               .first_value.values[0]
+
+  end_date = last_origin_info.field(datefields.join('|'))
+             .match_attribute(:point, 'end')
+             .first_value.values[0]
+
+  result.providedLabel = key_date if key_date
+  result.begin = begin_date if begin_date
+  result.end = end_date if end_date
+
+  if key_date || begin_date || end_date
+    result
+  else
+    []
+  end
+end
+
+# From the crosswalk document:
+#
+#   Grab all <relatedItem> values (including the collection name, which is the
+#   last <relatedItem>) and create separate values as in option one. Work from
+#   last to first tags. But first, strip all periods from the ends of all values
+#   and then put them together again with ". " space between the values to
+#   ensure we aren't doubling up on periods at the end of each value.
+#
+build_relation = lambda do |record|
+  titles = record['relatedItem'].field('titleInfo', 'title').values
+  if titles.empty?
+    []
+  else
+    titles.reverse.join('. ').gsub(/\.+/, '.')
+  end
+end
+
+# From the crosswalk document:
+#
+#   the last <titleInfo ... usage="primary"> <title>[value]</title> </titleInfo>
+#   (If <nonSort> appears nested within <titleInfo> grab that value and prepend
+#   it to <title>.)  OR (if there is no "primary" attribute) the first
+#   <titleInfo><title>
+#
+build_title = lambda do |record|
+  result = []
+
+  primary_title = record['titleInfo'].match_attribute(:usage, 'primary')
+                  .last_value
+
+  if primary_title.values.empty?
+    # If there's no primary title, just take what we can get.
+    title = record['titleInfo'].field('title')
+    result = title.values[0] unless title.values.empty?
+  else
+    # If there's a primary title, look for a nonSort and prepend it as needed.
+    result = primary_title.field('title').values[0]
+    non_sort = primary_title.field('nonSort')
+
+    result = non_sort.values[0] + result unless non_sort.values.empty?
+  end
+
+  result
+end
+
+Krikri::Mapper.define(:nypl_mods,
+                      parser: Krikri::ModsParser,
+                      parser_args: '//mods') do
+  provider class: DPLA::MAP::Agent do
+    uri 'http://dp.la/api/contributor/nypl'
+    label 'The New York Public Library'
+  end
+
+  dataProvider class: DPLA::MAP::Agent do
+    providedLabel record.field('location', 'physicalLocation')
+                   .match_attribute(:type, 'division')
+                   .reject_attribute(:authority, 'marcorg')
+                   .map { |i|
+                     i.value.strip.gsub(/\.+$/, '') +
+                                         '. The New York Public Library'
+                   }
+  end
+
+  isShownAt class: DPLA::MAP::WebResource do
+    uri record.field('extension', 'capture', 'itemLink')
+  end
+
+  preview class: DPLA::MAP::WebResource do
```

Comment:
I think this is wrong, based on [this `nypl_identify_object` enrichment module for the old system](https://github.com/dpla/ingestion/blob/develop/lib/akamod/nypl_identify_object.py). This should be the `imageID` element that exists at the level of a capture record, and last URI parameter should be `t=t` instead of `t=b`.